In [40]:
import numpy as np
import pandas as pd
from tqdm import tqdm
import matplotlib.pyplot as plt

## Importing data

In [2]:
PATH_TO_DATA1 = 'hackathon_data/avk_hackathon_data_account_x_balance.csv'
PATH_TO_DATA2 = 'hackathon_data/avk_hackathon_data_party_products.csv'
PATH_TO_DATA3 = 'hackathon_data/avk_hackathon_data_party_x_socdem.csv'
PATH_TO_DATA4 = 'hackathon_data/avk_hackathon_data_story_logs.csv'
PATH_TO_DATA5 = 'hackathon_data/avk_hackathon_data_story_texts.csv'
PATH_TO_DATA6 = 'hackathon_data/avk_hackathon_data_transactions.csv'

import json 
f = open('party_rk_top.json')
party_rk_top = json.load(f)

In [4]:
df1 = pd.read_csv(PATH_TO_DATA1)
df1 = df1[df1.party_rk.isin(party_rk_top)]
df1.head()


party_rk  account_rk  prev_month   cur_month  balance_chng
2      33941       33065  2018-12-31  2019-01-31      -15000.0
3       9336        8611  2018-12-31  2019-01-31      -50000.0
10       718         550  2018-12-31  2019-01-31       45000.0
12     11063       53925  2018-12-31  2019-01-31           0.0
18     40618       57186  2018-12-31  2019-01-31      -90000.0

In [6]:
df2 = pd.read_csv(PATH_TO_DATA2)
df2 = df2[df2.party_rk.isin(party_rk_top)]
df2.head()

party_rk  product1  product2  product3  product4  product5  product6  \
4      47253         0         0         0         1         1         0   
5      84832         0         0         0         1         1         0   
19     60144         0         0         0         1         0         0   
22     90259         0         0         0         1         1         0   
25     64148         0         1         0         0         1         0   

    product7  
4          0  
5          0  
19         0  
22         0  
25         0

In [7]:
df3 = pd.read_csv(PATH_TO_DATA3)
df3 = df3[df3.party_rk.isin(party_rk_top)]
df3.head()

party_rk gender_cd   age marital_status_desc  children_cnt  region_flg
3      88238         F  35.0                 NaN             0           0
7      24352         M  35.0       Женат/замужем             1           0
8        812         M  75.0                 NaN             4           1
13     89849         M  35.0       Женат/замужем             0           0
14     12883         M  45.0       Женат/замужем             0           1

In [11]:
df4 = pd.read_csv(PATH_TO_DATA4)
df4.date_time = pd.to_datetime(df4.date_time)
df4 = df4[df4.party_rk.isin(party_rk_top)]
df4.head()

party_rk           date_time  story_id  category event
9      48382 2019-12-10 12:06:47      3840        39  like
10     85548 2019-12-01 13:27:43      3846        39  like
13     46954 2019-12-01 14:08:50      3840        39  like
16     46954 2019-12-01 14:09:20      3857        39  like
22     74596 2019-12-01 14:38:45      3840        39  like

In [16]:
df5 = pd.read_csv(PATH_TO_DATA5)
df5.head()

story_id                              name  \
0       794  Профессия: свадебный организатор   
1       821       Бизнес: магазин электроники   
2       824               Как у них: Сингапур   
3       839                       Студии йоги   
4       848       Как получать больше кэшбэка   

                                          story_text  
0  ["","Как стать свадебным организатором","Это Л...  
1  ["","Как открыть интернет-магазин электроники"...  
2  ["","Жизнь в Сингапуре","Это Славина. Вместе с...  
3  ["","Студии йоги","7 московских студий, где мо...  
4  ["","Как получать больше кэшбэка","Подключите ...

In [17]:
df6 = pd.read_csv(PATH_TO_DATA6)
df6.transaction_dttm = pd.to_datetime(df6.transaction_dttm)
df6 = df6[df6.party_rk.isin(party_rk_top)]
df6.head()

party_rk  account_rk  financial_account_type_cd transaction_dttm  \
0      20337       19666                          1       2019-01-01   
2      24789       23517                          2       2019-01-01   
3      57970       64838                          2       2019-01-01   
12     56812       63010                          2       2019-01-01   
23     21861       20802                          2       2019-01-01   

   transaction_type_desc  transaction_amt_rur  merchant_rk  merchant_type  \
0                Покупка                84.00      88676.0          348.0   
2                Покупка               701.44     830014.0          291.0   
3                Покупка              6203.70     363834.0          278.0   
12               Покупка              1695.00     988289.0          346.0   
23       Снятие наличных              5000.00     439054.0          380.0   

    merchant_group_rk       category  
0                 NaN       Сувениры  
2                 NaN   Супермаркеты  
3               454.0     Дом/Ремонт  
12              100.0  Разные товары  
23                NaN       Наличные

## Trying to compare category of stories and category of transactions

In [28]:
dict_of_categories = {2: ['Финансовые услуги'],

4: ['Ж/д билеты', 'Авиабилеты', 'Турагентства', 'Отели', 'Duty Free'], 

20: ['Авиабилеты', 'Турагентства', 'Отели', 'Duty Free'], 
22: ['Авиабилеты', 'Турагентства', 'Отели', 'Duty Free'], 
23: ['Авиабилеты', 'Турагентства', 'Отели', 'Duty Free'], 

21: ['Авиабилеты'], 

27: ['Автоуслуги', 'Госсборы', 'Транспорт', 'Аренда авто'],

30: ['Искусство'],

31: ['Музыка'],

33: ['Супермаркеты'],

38: ['Спорттовары'],

41: ['Авиабилеты', 'Турагентства', 'Отели', 'Duty Free', 'Сувениры'], 

43: ['Ж/д билеты', 'Авиабилеты', 'Турагентства', 'Отели', 'Duty Free', 'Сувениры'], 

44: ['Кино', 'Развлечения'], 

49: ['Искусство', 'Развлечения'],

51: ['Рестораны', 'Фаст Фуд'],

52: ['Книги'],

55: ['Развлечения'],

64: ['Авиабилеты', 'Турагентства', 'Отели', 'Duty Free', 'Сувениры']}


### some preprocessing

In [29]:
df_liked_fav = df4[df4.event.isin(['like', 'favorite'])]


In [37]:
df_liked_fav.iloc[0].loc['date_time']

Timestamp('2019-12-10 12:06:47')

In [38]:
int(time.mktime(df_liked_fav.iloc[0].loc['date_time'].timetuple()))

1575968807

In [33]:
import datetime
import time

In [44]:
df_liked_fav

party_rk           date_time  story_id  category     event  \
9         48382 2019-12-10 12:06:47      3840        39      like   
10        85548 2019-12-01 13:27:43      3846        39      like   
13        46954 2019-12-01 14:08:50      3840        39      like   
16        46954 2019-12-01 14:09:20      3857        39      like   
22        74596 2019-12-01 14:38:45      3840        39      like   
...         ...                 ...       ...       ...       ...   
13806     18049 2019-08-11 08:16:49      2827        43      like   
13807      6350 2019-08-11 14:19:39      2784        51      like   
13817     10195 2019-12-31 23:32:50      4117        71  favorite   
13818     10195 2019-12-31 23:33:32      4120        39      like   
13819     10195 2019-12-31 23:32:43      4117        71      like   

       categories_for_recommend last_day_for_recommendation  
9                            39         2019-12-10 12:06:47  
10                           39         2019-12-01 13:27:43  
13                           39         2019-12-01 14:08:50  
16                           39         2019-12-01 14:09:20  
22                           39         2019-12-01 14:38:45  
...                         ...                         ...  
13806                        43         2019-08-11 08:16:49  
13807                        51         2019-08-11 14:19:39  
13817                        71         2019-12-31 23:32:50  
13818                        39         2019-12-31 23:33:32  
13819                        71         2019-12-31 23:32:43  

[4334 rows x 7 columns]

## Recommendation code

the structure is pretty simple: if person liked/favorited some stories, he will probably make a transaction in category connected to the story category

In [59]:
df_liked_fav['categories_for_recommend'] = df_liked_fav['category']
df_liked_fav['last_day_for_recommendation'] = df_liked_fav['date_time']
df_liked_fav['categories_for_recommend'] = df_liked_fav['categories_for_recommend'].astype(object)
for index, row in tqdm(df_liked_fav.iterrows()):
    if df_liked_fav.loc[index, 'category'] in list(dict_of_categories.keys()):
#         print (df_liked_fav.loc[index, 'category'] )
#         print(dict_of_categories[df_liked_fav.loc[index, 'category']])
        df_liked_fav.at[index, 'categories_for_recommend'] = list(dict_of_categories[df_liked_fav.loc[index, 'category']])
        df_liked_fav.loc[index, 'last_day_for_recommendation'] = datetime.datetime.fromtimestamp(int(time.mktime(df_liked_fav.loc[index, 'date_time'].timetuple())) + 2592000)
    else:
        df_liked_fav.at[index, 'categories_for_recommend'] = np.nan
        df_liked_fav.loc[index, 'last_day_for_recommendation'] = np.nan

C:\Users\Miron\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
C:\Users\Miron\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
C:\Users\Miron\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in

In [60]:
df_liked_fav

party_rk           date_time  story_id  category     event  \
9         48382 2019-12-10 12:06:47      3840        39      like   
10        85548 2019-12-01 13:27:43      3846        39      like   
13        46954 2019-12-01 14:08:50      3840        39      like   
16        46954 2019-12-01 14:09:20      3857        39      like   
22        74596 2019-12-01 14:38:45      3840        39      like   
...         ...                 ...       ...       ...       ...   
13806     18049 2019-08-11 08:16:49      2827        43      like   
13807      6350 2019-08-11 14:19:39      2784        51      like   
13817     10195 2019-12-31 23:32:50      4117        71  favorite   
13818     10195 2019-12-31 23:33:32      4120        39      like   
13819     10195 2019-12-31 23:32:43      4117        71      like   

                                categories_for_recommend  \
9                                                    NaN   
10                                                   NaN   
13                                                   NaN   
16                                                   NaN   
22                                                   NaN   
...                                                  ...   
13806  [Ж/д билеты, Авиабилеты, Турагентства, Отели, ...   
13807                              [Рестораны, Фаст Фуд]   
13817                                                NaN   
13818                                                NaN   
13819                                                NaN   

      last_day_for_recommendation  
9                             NaT  
10                            NaT  
13                            NaT  
16                            NaT  
22                            NaT  
...                           ...  
13806         2019-09-10 08:16:49  
13807         2019-09-10 14:19:39  
13817                         NaT  
13818                         NaT  
13819                         NaT  

[4334 rows x 7 columns]

Let's test it on data that we have

In [64]:
without_nan_df_liked_fav = df_liked_fav.dropna()
without_nan_df_liked_fav

party_rk           date_time  story_id  category     event  \
54         8049 2019-12-02 18:33:44      3837        51  favorite   
149       88817 2019-12-16 11:07:28      3931        49  favorite   
154        3645 2019-12-20 09:22:14      4019        43      like   
177       31024 2019-12-01 08:37:31      3627        20      like   
192       34650 2019-12-30 11:48:48      4077        43      like   
...         ...                 ...       ...       ...       ...   
13783     81798 2019-08-12 02:00:34      2830        51      like   
13801     93847 2019-05-23 14:46:07      2562        51  favorite   
13802     66389 2019-05-23 15:07:53      2609        51      like   
13806     18049 2019-08-11 08:16:49      2827        43      like   
13807      6350 2019-08-11 14:19:39      2784        51      like   

                                categories_for_recommend  \
54                                 [Рестораны, Фаст Фуд]   
149                             [Искусство, Развлечения]   
154    [Ж/д билеты, Авиабилеты, Турагентства, Отели, ...   
177         [Авиабилеты, Турагентства, Отели, Duty Free]   
192    [Ж/д билеты, Авиабилеты, Турагентства, Отели, ...   
...                                                  ...   
13783                              [Рестораны, Фаст Фуд]   
13801                              [Рестораны, Фаст Фуд]   
13802                              [Рестораны, Фаст Фуд]   
13806  [Ж/д билеты, Авиабилеты, Турагентства, Отели, ...   
13807                              [Рестораны, Фаст Фуд]   

      last_day_for_recommendation  
54            2020-01-01 18:33:44  
149           2020-01-15 11:07:28  
154           2020-01-19 09:22:14  
177           2019-12-31 08:37:31  
192           2020-01-29 11:48:48  
...                           ...  
13783         2019-09-11 02:00:34  
13801         2019-06-22 14:46:07  
13802         2019-06-22 15:07:53  
13806         2019-09-10 08:16:49  
13807         2019-09-10 14:19:39  

[1303 rows x 7 columns]

In [71]:
from collections import Counter
correct_rec = 0
incorrect_rec = 0
transactions_correct_rec = 0
for index, row in tqdm(without_nan_df_liked_fav.iterrows()):
    buffer_df = df6[df6.party_rk == without_nan_df_liked_fav.loc[index, 'party_rk']]
    buffer_df = buffer_df[buffer_df.transaction_dttm > without_nan_df_liked_fav.loc[index, 'date_time']]
    buffer_df = buffer_df[buffer_df.transaction_dttm < without_nan_df_liked_fav.loc[index, 'last_day_for_recommendation']]
    counter_dict = Counter(list(buffer_df.category))
    for rec_cat in without_nan_df_liked_fav.loc[index, 'categories_for_recommend']:
        if rec_cat in list(buffer_df.category):
            correct_rec+=1
        if rec_cat in list(counter_dict.keys()):
            transactions_correct_rec += counter_dict[rec_cat]
        else:
            incorrect_rec += 1

1303it [00:07, 181.77it/s]


In [72]:
print(correct_rec, incorrect_rec, transactions_correct_rec)

1031 2608 4688


Results is showing us that we have recommended right category for 1031 times. And people have made 4688 transactions on them.

2608 were not right, but in our opinion, if people will be supported with some Tinkoff's offers in these categories, amount will be less. It can be tested only in real life, not in expirements. For example, AB-test will help